### DISCUSSION TITLE
How embed .NET in native application?

### DISCUSSION THREAD

**Discussion by jppgmx** [dotnet/runtime#79309] [Q&A] [answered]
Currently, in certain programming languages ​​such as Java and Python it can be embedded in a native application. For example, JNI can create a VM for Java and the Python interpreter is initialized with `Py_Initialize()`.

I found a way to embed .NET in a native application, with C++/CLI, but I saw that in the documentation for migrating C++/CLI to pure .NET, this feature is only for Windows.

So I looked for other ways and found three possibilities for cross platform:

- Mono;
- hostfxr library;
- coreclr library.

Mono I found a bit complicated, hostfxr I found good, but it has a difficult initialization structure. But CoreCLR I found it perfect, simply loading coreclr.dll, calling its functions and has function to shut down.

But also, I was undecided, among those above, which one is better to embed, where I can find it and a small example. (I already searched here and didn't find anything related to this.)

**Comment by vitek-karas** [comment, marked as answer]
The recommended way is the `hostfxr` APIs.
https://learn.microsoft.com/en-us/dotnet/core/tutorials/netcore-hosting
https://github.com/dotnet/samples/tree/main/core/hosting

Curious: What are the initialization difficulties you ran into using it?

Please do not use coreclr directly - we don't consider it a publicly supported interface and may make breaking changes.

Also note that .NET 6+ doesn't really implement the "shutdown" - it will do some things, but it will not unload the runtime from the process - and it's not valid to load a new runtime later on into the same process.

**Reply by jppgmx** [reply]
> What are the initialization difficulties you ran into using it?

Basically, if `hostfxr` is found on different platforms, I already use Windows to test it there, but I don't know if it will work the same on other platforms, but I can adapt it using CI anyway in some tests.

Since `hostfxr` does not implement shutdown, what would its lifecycle be when started? At what point does it end?
Thanks in advance for the answer.

**Reply by vitek-karas** [reply]
`hostfxr` works on all platforms - the code should be basically the same. Look at the sample: https://github.com/dotnet/samples/tree/main/core/hosting you should be able to run that on all platforms as-is.

It's not that `hostfxr` doesn't implement shutdown, the runtime itself doesn't really implement shutdown. I know it has the `coreclr_shutdown` method, but that method doesn't fully shutdown the runtime. The runtime code will still be loaded into the process. It will probably free some memory, but likely not all.
The current design is that once the runtime is loaded into the process, it stays there till the process dies.
